### PR TITLE
Support graceful shutdown for Broker

### PR DESCRIPTION
**Pull request by 315157973**
### Motivation
Now the following problems exist when the broker is closing:
1. When there are many topics on the Broker that is being closed, there will be a large number of concurrent lookup requests, which will eventually cause the Lookup to time out.
2. When `broker.close()` is running, it directly sets the Topic state to terminated without waiting for the client connection to be closed, resulting in an exception on the client side and obvious perception.

### Modifications
1. Add a REST API for graceful shutdown
2. The rate of unload per second can be controlled in the API
3. When closing the broker, we can wait for the client connection to close, thereby making the broker switching smoother

### Documentation
- [ x ] `no-need-doc`

### CODE REVIEW

**Review by Jason918 — COMMENTED** [review]
Can we make this a default behavior for `broker.closeAsync()`?
After we called this `shutdown` api, the broker process still exists, right?

**Comment by 315157973** [comment]
> Can we make this a default behavior for `broker.closeAsync()`? After we called this `shutdown` api, the broker process still exists, right?

`broker.closeAsync()` already has the ability to close gracefully, but it cannot limit the rate of unload, so I added a new API

The broker process should exit, and I should call `PulsarService.close` in my code

**Comment by mattisonchao** [comment]
Hi, @315157973

one question:
- Since we call "broker.closeAsync()" asynchronously, how does the user know if the broker is really closed? because user always gets 200 after ``unloadNamespaceBundlesGracefully``.

I mean we could enhance the admin side to check broker if is alive after calling ``shutDownBrokerGracefully``. 

Thanks for your answer.

**Comment by 315157973** [comment]
> Hi, @315157973
> 
> one question:
> 
> * Since we call `broker.closeAsync()` asynchronously, how does the user know if the broker is really closed? because the user always gets 200 after `unloadNamespaceBundlesGracefully`.
> 
> I mean we could enhance the admin side to check broker if is alive after calling `shutDownBrokerGracefully`.
> 
> Thanks for your answer.

If done, the process will exit.

**Review by Jason918 — APPROVED** [review]
LGTM

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java [hunk 1]
```diff
@@ -446,5 +447,27 @@ private synchronized void deleteDynamicConfigurationOnZk(String configName) {
...
+    @POST
+    @Path("/shutdown")
+    @ApiOperation(value =
+            "Shutdown broker gracefully.")
+    @ApiResponses(value = {
+            @ApiResponse(code = 204, message = "Execute shutdown command successfully"),
+            @ApiResponse(code = 403, message = "You don't have admin permission to update service-configuration"),
+            @ApiResponse(code = 500, message = "Internal server error")})
+    public void shutDownBrokerGracefully(
+            @QueryParam("maxConcurrentUnloadPerSec") int maxConcurrentUnloadPerSec,
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:460]: Better to add a description here, if the value absent(value=0) means no concurrent limitation.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java [hunk 2]
```diff
@@ -446,5 +447,27 @@ private synchronized void deleteDynamicConfigurationOnZk(String configName) {
...
+            @QueryParam("forcedTerminateTopic") @DefaultValue("true") boolean forcedTerminateTopic
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:461]: Does the default value is `true` if without `?forcedTerminateTopic=true`?

**Review by codelipenghui — COMMENTED** [review]
Please also add the CLI.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java (hunk 2 above)
Comment (315157973) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokersBase.java:461]: Yes

**Review by codelipenghui — APPROVED** [review]

**Comment by sagar-upswing** [comment]
> `broker.closeAsync()` already has the ability to close gracefully, but it cannot limit the rate of unload, so I added a new API

@315157973 hie,
Would you be able to share some more details/doc on how it actually gracefully shuts down, cause I am facing some issues w/ restarts under k8s?

**Pull request merged**